File: src/engine/adapters/clickhouse/read.ts
import type { ColumnMeta } from '../../../shared/domain/tree';
import type { SortSpec } from '../../../shared/protocol/data-ops';
import {
  type ColumnDescriptor,
  createTabularPageBuilder,
  type PagePosition,
  type TabularPage,
  type TypeClass,
} from '../../../shared/protocol/page';
import type { OpCtx, ReadRequest } from '../adapter';
import { AdapterError } from '../errors';
import { buildOrderBy } from '../sql-text';
import type { ReadTarget } from './catalog';
import type { ClickHouseHandle } from './client';
import { runCatalogQuery, streamQuery, type TrackQuery } from './query';

// F28/D29: ClickHouse quotes identifiers with backticks — the same character its own
// `create_table_query` output uses (F14's example), so this is what the app emits back.
export function quoteIdent(name: string): string {
  return `\`${name.replace(/`/g, '``')}\``;
}

// D17: unwraps Nullable(...) and LowCardinality(...) — recursively, in either nesting order,
// since `LowCardinality(Nullable(String))` and `Nullable(LowCardinality(String))` are both real,
// common column types (F24) — before classifying the inner type name.
export function unwrapType(declared: string): { inner: string; nullable: boolean } {
  let inner = declared.trim();
  let nullable = false;
  for (;;) {
    const nullableMatch = /^Nullable\((.*)\)$/s.exec(inner);
    if (nullableMatch?.[1] !== undefined) {
      nullable = true;
      inner = nullableMatch[1].trim();
      continue;
    }
    const lowCardMatch = /^LowCardinality\((.*)\)$/s.exec(inner);
    if (lowCardMatch?.[1] !== undefined) {
      inner = lowCardMatch[1].trim();
      continue;
    }
    break;
  }
  return { inner, nullable };
}

function baseTypeName(inner: string): string {
  const match = /^[A-Za-z][A-Za-z0-9]*/.exec(inner);
  return match ? match[0] : inner;
}

const TEXT_TYPES = new Set(['String', 'FixedString', 'UUID', 'IPv4', 'IPv6', 'Enum8', 'Enum16']);
const TEMPORAL_TYPES = new Set(['Date', 'Date32', 'DateTime', 'DateTime64', 'Time', 'Time64']);
const JSON_TYPES = new Set([
  'JSON',
  'Dynamic',
  'Variant',
  'Array',
  'Tuple',
  'Map',
  'Nested',
  'Point',
  'Ring',
  'Polygon',
  'MultiPolygon',
]);

// D17: 'String' -> text, never 'binary' — ClickHouse's own docs describe it as replacing
// VARCHAR, BLOB and CLOB alike (F24), and there is no per-column signal to tell them apart;
// guessing binary would put every text column behind the cell editor's hex pane. The composite
// and semi-structured types -> 'json': ClickHouse renders Array/Tuple/Map/the geo types as JSON
// arrays/objects in this adapter's own wire format (D16), so the cell editor's JSON pane is
// already the right rendering for them (the same "sugar" judgement P35 D21 made for three types,
// applied to five families here). 'other' is the honest answer for AggregateFunction,
// SimpleAggregateFunction, Nothing, Interval and anything unrecognised — never guessed at.
export function typeClassFor(declared: string): TypeClass {
  const { inner } = unwrapType(declared);
  const name = baseTypeName(inner);
  if (name === 'Bool') return 'boolean';
  if (/^(Int|UInt|Float|Decimal)/.test(name)) return 'number';
  if (TEMPORAL_TYPES.has(name)) return 'temporal';
  if (TEXT_TYPES.has(name)) return 'text';
  if (JSON_TYPES.has(name)) return 'json';
  return 'other';
}

// App-generated integers only (pageSize+1, an offset already validated at the port boundary) —
// inlined rather than bound, mirroring every other SQL adapter's identical discipline.
function safeInt(value: number, label: string): number {
  if (!Number.isSafeInteger(value) || value < 0) {
    throw new AdapterError('E_QUERY', `invalid ${label}: ${value}`);
  }
  return value;
}

function resolveProjection(target: ReadTarget, requested: string[] | null): ColumnMeta[] {
  if (requested === null) return target.columns;
  const byName = new Map(target.columns.map((c) => [c.name, c]));
  const resolved: ColumnMeta[] = [];
  for (const name of new Set(requested)) {
    const col = byName.get(name);
    if (!col) throw new AdapterError('E_NOT_FOUND', `unknown column in projection: ${name}`);
    resolved.push(col);
  }
  resolved.sort((a, b) => a.position - b.position);
  return resolved;
}

// D21: a requested sort is honoured as given; with none, the table's own sorting key is used
// verbatim (F31 — the one ORDER BY ClickHouse can serve straight out of the parts' existing
// order), rather than re-deriving a column list from is_in_sorting_key flags, which would be a
// different, slower order the engine could no longer read in place. Still not a *total* order:
// duplicate sorting-key values tie arbitrarily (§5.1's own note).
function computeOrderBySql(sort: SortSpec | null, target: ReadTarget): string {
  if (sort?.kind === 'text') return sort.text;
  if (sort?.kind === 'structured' && sort.terms.length > 0) {
    const byName = new Set(target.columns.map((c) => c.name));
    for (const t of sort.terms) {
      if (!byName.has(t.column))
        throw new AdapterError('E_NOT_FOUND', `unknown column in sort: ${t.column}`);
    }
    return buildOrderBy(sort.terms, quoteIdent);
  }
  return target.sortingKey.trim();
}

const NO_KEYSET_MESSAGE =
  'keyset pagination is unavailable for ClickHouse: a MergeTree PRIMARY KEY is a sparse index, ' +
  'not a unique key, so there is no total order to build a keyset cursor on — use an offset cursor.';

// D20: caps.pagination is 'offset' — every page is offset-strategy, unconditionally. A cursor in
// 'after'/'before' mode is refused outright rather than silently falling back, since silently
// switching strategies out from under a caller expecting keyset semantics would be its own kind
// of dishonesty.
export async function readPage(
  h: ClickHouseHandle,
  ctx: OpCtx,
  target: ReadTarget,
  req: Omit<ReadRequest, 'path'>,
  track: TrackQuery,
  nextQueryId: () => string,
): Promise<TabularPage> {
  if (req.cursor.mode !== 'offset') {
    throw new AdapterError('E_UNSUPPORTED', NO_KEYSET_MESSAGE);
  }

  const projectedColumns = resolveProjection(target, req.projection);
  const columns: ColumnDescriptor[] = projectedColumns.map((c) => ({
    name: c.name,
    dataType: c.dataType,
    typeClass: typeClassFor(c.dataType),
    nullable: c.nullable,
    isPrimaryKey: c.isPrimaryKey,
  }));

  const relationSql = `${quoteIdent(target.qualifiedName.schema)}.${quoteIdent(target.qualifiedName.table)}`;
  const selectList = projectedColumns.map((c) => quoteIdent(c.name)).join(', ');
  const whereSql = req.filter && req.filter.trim() !== '' ? `WHERE (${req.filter})` : '';
  const orderBySql = computeOrderBySql(req.sort, target);
  const limit = safeInt(req.pageSize + 1, 'page size');
  const offset = safeInt(req.cursor.offset, 'offset');

  const sql = [
    `SELECT ${selectList}`,
    `FROM ${relationSql}`,
    whereSql,
    orderBySql ? `ORDER BY ${orderBySql}` : '',
    `LIMIT ${limit} OFFSET ${offset}`,
  ]
    .filter(Boolean)
    .join('\n');

  ctx.setCommand(sql);
  const builder = createTabularPageBuilder(columns);
  let rowCount = 0;
  let hasMore = false;
  await streamQuery(
    h,
    ctx,
    sql,
    { queryId: nextQueryId() },
    track,
    () => {},
    (values) => {
      if (rowCount >= req.pageSize) {
        hasMore = true;
        return;
      }
      builder.appendRow(values);
      rowCount++;
    },
  );

  const position: PagePosition = {
    offset: req.cursor.offset,
    pageSize: req.pageSize,
    hasMore,
    nextToken: null,
    prevToken: null,
    strategy: 'offset',
  };
  return builder.finish(position);
}

export async function countRows(
  h: ClickHouseHandle,
  ctx: OpCtx,
  target: Pick<ReadTarget, 'qualifiedName'>,
  filter: string | null,
  track: TrackQuery,
  nextQueryId: () => string,
): Promise<{ value: number; exact: boolean }> {
  const relationSql = `${quoteIdent(target.qualifiedName.schema)}.${quoteIdent(target.qualifiedName.table)}`;
  const whereSql = filter && filter.trim() !== '' ? `WHERE (${filter})` : '';
  const sql = [`SELECT count() AS n`, `FROM ${relationSql}`, whereSql].filter(Boolean).join('\n');

  const rows = await runCatalogQuery<{ n: string }>(h, ctx, sql, { queryId: nextQueryId() }, track);
  const raw = rows[0]?.n;
  const value = Number(raw ?? '0');
  if (!Number.isFinite(value)) {
    throw new AdapterError('E_QUERY', `count returned a non-numeric result: ${String(raw)}`);
  }
  return { value, exact: true };
}

File: apps/kira-studio/internal/adapters/clickhouse/read.go
package clickhouse

import (
	"context"
	"regexp"
	"strings"

	"github.com/kirathecat/kira-studio/apps/kira-studio/internal/adapters"
	"github.com/kirathecat/kira-studio/apps/kira-studio/internal/page"
	"github.com/kirathecat/kira-studio/apps/kira-studio/internal/storage/model"
)

// quoteIdent is read.ts's own — F28/D29: ClickHouse quotes identifiers with backticks, the same
// character its own create_table_query output uses.
//
// P2 R1: ClickHouse's lexer reads a backtick-quoted identifier with the same backslash-escape
// rules as a string literal, not just SQL-style doubled-quote escaping — a raw backslash must be
// doubled too, or a name ending in one (or containing "\`") lets the following character,
// including the identifier's own closing backtick, be consumed as an escape sequence instead of
// terminating the identifier, corrupting the query or letting subsequent text run as SQL.
func quoteIdent(name string) string {
	if strings.ContainsRune(name, '\x00') {
		panic(adapters.New(adapters.CodeQuery, "identifier contains a NUL byte", nil))
	}
	escaped := strings.ReplaceAll(name, "\\", "\\\\")
	escaped = strings.ReplaceAll(escaped, "`", "``")
	return "`" + escaped + "`"
}

var (
	nullableRE      = regexp.MustCompile(`(?s)^Nullable\((.*)\)$`)
	lowCardinalityR = regexp.MustCompile(`(?s)^LowCardinality\((.*)\)$`)
	baseTypeNameRE  = regexp.MustCompile(`^[A-Za-z][A-Za-z0-9]*`)
)

// unwrapType is read.ts's own — D17: unwraps Nullable(...) and LowCardinality(...) recursively, in
// either nesting order, since LowCardinality(Nullable(String)) and Nullable(LowCardinality(String))
// are both real, common column types (F24).
func unwrapType(declared string) (inner string, nullable bool) {
	inner = strings.TrimSpace(declared)
	for {
		if m := nullableRE.FindStringSubmatch(inner); m != nil {
			nullable = true
			inner = strings.TrimSpace(m[1])
			continue
		}
		if m := lowCardinalityR.FindStringSubmatch(inner); m != nil {
			inner = strings.TrimSpace(m[1])
			continue
		}
		break
	}
	return inner, nullable
}

func baseTypeName(inner string) string {
	if m := baseTypeNameRE.FindString(inner); m != "" {
		return m
	}
	return inner
}

var (
	numberTypeRE  = regexp.MustCompile(`^(Int|UInt|Float|Decimal)`)
	textTypes     = map[string]bool{"String": true, "FixedString": true, "UUID": true, "IPv4": true, "IPv6": true, "Enum8": true, "Enum16": true}
	temporalTypes = map[string]bool{"Date": true, "Date32": true, "DateTime": true, "DateTime64": true, "Time": true, "Time64": true}
	jsonTypes     = map[string]bool{
		"JSON": true, "Dynamic": true, "Variant": true, "Array": true, "Tuple": true, "Map": true,
		"Nested": true, "Point": true, "Ring": true, "Polygon": true, "MultiPolygon": true,
	}
)

// typeClassFor is read.ts's own — D17: 'String' -> text, never 'binary'; the composite and
// semi-structured types -> 'json'; 'other' for AggregateFunction/SimpleAggregateFunction/Nothing/
// Interval/anything unrecognised, never guessed at.
func typeClassFor(declared string) page.TypeClass {
	inner, _ := unwrapType(declared)
	name := baseTypeName(inner)
	switch {
	case name == "Bool":
		return page.TypeClassBoolean
	case numberTypeRE.MatchString(name):
		return page.TypeClassNumber
	case temporalTypes[name]:
		return page.TypeClassTemporal
	case textTypes[name]:
		return page.TypeClassText
	case jsonTypes[name]:
		return page.TypeClassJSON
	default:
		return page.TypeClassOther
	}
}

// computeOrderBySql is read.ts's own — D21: a requested sort is honoured as given; with none, the
// table's own sorting key is used verbatim (F31), rather than re-deriving a column list, which
// would be a different, slower order the engine could no longer read in place.
func computeOrderBySql(sort *model.SortSpec, target ReadTarget) (string, error) {
	if sort != nil && sort.Kind == "text" {
		return sort.Text, nil
	}
	if sort != nil && sort.Kind == "structured" && len(sort.Terms) > 0 {
		byName := make(map[string]bool, len(target.Columns))
		for _, c := range target.Columns {
			byName[c.Name] = true
		}
		terms := make([]adapters.OrderTerm, len(sort.Terms))
		for i, t := range sort.Terms {
			if !byName[t.Column] {
				return "", adapters.New(adapters.CodeNotFound, "unknown column in sort: "+t.Column, nil)
			}
			terms[i] = adapters.OrderTerm{Column: t.Column, Direction: t.Direction}
		}
		return adapters.BuildOrderBy(terms, quoteIdent), nil
	}
	return strings.TrimSpace(target.SortingKey), nil
}

const noKeysetMessage = "keyset pagination is unavailable for ClickHouse: a MergeTree PRIMARY KEY is a sparse index, " +
	"not a unique key, so there is no total order to build a keyset cursor on — use an offset cursor."

// readReq is adapter.ts's ReadRequest minus Path.
type readReq struct {
	Projection []string
	Filter     *string
	Sort       *model.SortSpec
	PageSize   int
	Cursor     model.PageCursor
}

// readPage is read.ts's own — D20: caps.Pagination is offset, unconditionally; a cursor in
// after/before mode is refused outright rather than silently falling back to offset, since a
// silent strategy switch out from under a caller expecting keyset semantics would be its own kind
// of dishonesty.
func readPage(ctx context.Context, h *Handle, queryID string, op *adapters.OpCtx, track TrackQuery, target ReadTarget, req readReq) (page.TabularPage, error) {
	if req.Cursor.Mode != "offset" {
		return page.TabularPage{}, adapters.New(adapters.CodeUnsupported, noKeysetMessage, nil)
	}

	projectedColumns, err := adapters.ResolveProjection(target.Columns, req.Projection)
	if err != nil {
		return page.TabularPage{}, err
	}
	columns := make([]page.ColumnDescriptor, len(projectedColumns))
	for i, c := range projectedColumns {
		columns[i] = page.ColumnDescriptor{
			Name: c.Name, DataType: c.DataType, TypeClass: typeClassFor(c.DataType),
			Nullable: c.Nullable, IsPrimaryKey: c.IsPrimaryKey,
			Generated: target.GeneratedColumns[c.Name],
		}
	}

	relationSQL := quoteIdent(target.QualifiedName.Database) + "." + quoteIdent(target.QualifiedName.Table)
	selectNames := make([]string, len(projectedColumns))
	for i, c := range projectedColumns {
		selectNames[i] = quoteIdent(c.Name)
	}
	selectList := strings.Join(selectNames, ", ")
	whereSQL := adapters.WhereClause(req.Filter)
	orderBySQL, err := computeOrderBySql(req.Sort, target)
	if err != nil {
		return page.TabularPage{}, err
	}
	limit, err := adapters.SafeInt(req.PageSize+1, "page size")
	if err != nil {
		return page.TabularPage{}, err
	}
	offset, err := adapters.SafeInt(req.Cursor.Offset, "offset")
	if err != nil {
		return page.TabularPage{}, err
	}

	sqlParts := []string{"SELECT " + selectList, "FROM " + relationSQL}
	if whereSQL != "" {
		sqlParts = append(sqlParts, whereSQL)
	}
	if orderBySQL != "" {
		sqlParts = append(sqlParts, "ORDER BY "+orderBySQL)
	}
	sqlParts = append(sqlParts, "LIMIT "+itoaPositive(limit)+" OFFSET "+itoaPositive(offset))
	sql := strings.Join(sqlParts, "\n")

	op.SetCommand(sql)
	builder := page.NewTabularPageBuilder(columns)
	rowCount := 0
	hasMore := false
	err = StreamQuery(ctx, h, queryID, sql, op, track, func(names, types []string) {}, func(values []*string) {
		if rowCount >= req.PageSize {
			hasMore = true
			return
		}
		_ = builder.AppendRow(values)
		rowCount++
	})
	if err != nil {
		return page.TabularPage{}, err
	}

	offsetCopy := req.Cursor.Offset
	position := page.PagePosition{
		Offset: &offsetCopy, PageSize: req.PageSize, HasMore: hasMore,
		NextToken: nil, PrevToken: nil, Strategy: "offset",
	}
	return builder.Finish(position), nil
}

// countRows is read.ts's own.
func countRows(ctx context.Context, h *Handle, queryID string, op *adapters.OpCtx, track TrackQuery, target QualifiedName, filter *string) (adapters.CountResult, error) {
	relationSQL := quoteIdent(target.Database) + "." + quoteIdent(target.Table)
	sqlParts := []string{"SELECT count() AS n", "FROM " + relationSQL}
	if where := adapters.WhereClause(filter); where != "" {
		sqlParts = append(sqlParts, where)
	}
	sql := strings.Join(sqlParts, "\n")

	rows, err := RunCatalogQuery[struct {
		N string `json:"n"`
	}](ctx, h, queryID, sql, op, track, nil)
	if err != nil {
		return adapters.CountResult{}, err
	}
	n := "0"
	if len(rows) > 0 {
		n = rows[0].N
	}
	value, err := adapters.ParseCountValue(n)
	if err != nil {
		return adapters.CountResult{}, err
	}
	return adapters.CountResult{Value: value, Exact: true}, nil
}
